kani-compiler/src/codegen_boogie/context/boogie_ctx.rs

```diff
@@ -13,14 +29,303 @@ pub struct BoogieCtx<'tcx> {
+                if self.layout_of(typ).is_zst() {
+                    return None;
+                }
+                debug!(?lc, ?typ, "codegen_declare_variables");
+                let name = format!("{lc:?}");
+                let boogie_type = self.codegen_type(typ);
+                Some(Stmt::Decl { name, typ: boogie_type })
+            })
+            .collect();
+        decls
+    }
+
+    pub fn codegen_type(&self, ty: Ty<'tcx>) -> Type {
+        trace!(typ=?ty, "codegen_type");
+        match ty.kind() {
+            ty::Bool => Type::Bool,
+            ty::Int(_ity) => Type::Int, // TODO: use Bv
+            _ => todo!(),
+        }
+    }
+
+    pub fn codegen_body(&self, instance: Instance<'tcx>) -> Stmt {
+        let mir = self.tcx.instance_mir(instance.def);
+        let statements: Vec<Stmt> = reverse_postorder(mir)
+            .map(|(bb, bbd)| self.codegen_block(mir.local_decls(), bb, bbd))
+            .collect();
+        Stmt::Block { statements }
+    }
+
+    pub fn codegen_block(
+        &self,
+        local_decls: &LocalDecls<'tcx>,
+        bb: BasicBlock,
+        bbd: &BasicBlockData<'tcx>,
+    ) -> Stmt {
+        debug!(?bb, ?bbd, "codegen_block");
+        // the first statement should be labelled. if there is no statements, then the
+        // terminator should be labelled.
+        let statements = match bbd.statements.len() {
+            0 => {
+                let term = bbd.terminator();
+                let tcode = self.codegen_terminator(local_decls, term);
+                vec![tcode]
+            }
+            _ => {
+                let mut statements: Vec<Stmt> =
+                    bbd.statements.iter().map(|stmt| self.codegen_statement(stmt)).collect();
+
+                let term = self.codegen_terminator(local_decls, bbd.terminator());
+                statements.push(term);
+                statements
+            }
+        };
+        Stmt::Block { statements }
+    }
+
+    pub fn codegen_statement(&self, stmt: &Statement<'tcx>) -> Stmt {
+        match &stmt.kind {
+            StatementKind::Assign(box (place, rvalue)) => {
+                debug!(?place, ?rvalue, "codegen_statement");
+                let rv = self.codegen_rvalue(rvalue);
+                Stmt::Assignment { target: format!("{:?}", place.local), value: rv }
+            }
+            StatementKind::FakeRead(..)
+            | StatementKind::SetDiscriminant { .. }
+            | StatementKind::Deinit(..)
+            | StatementKind::StorageLive(..)
+            | StatementKind::StorageDead(..)
+            | StatementKind::Retag(..)
+            | StatementKind::PlaceMention(..)
+            | StatementKind::AscribeUserType(..)
+            | StatementKind::Coverage(..)
+            | StatementKind::Intrinsic(..)
+            | StatementKind::ConstEvalCounter
+            | StatementKind::Nop => todo!(),
+        }
+    }
+
+    pub fn codegen_rvalue(&self, rvalue: &Rvalue<'tcx>) -> Expr {
```

Comment:
I wonder if we should follow this pattern of returning expression in these functions. In the `goto` translation, we constantly add checks that also generate statements. Does Boogie have anything similar to `goto`'s `StatementExpression`?

Another alternative is to ensure the checks are added before codegen. We could either add these checks as MIR transformation pass or fix https://github.com/model-checking/kani/issues/1740.

Reply:
I don't think Boogie has something similar. Adding the checks as a MIR transformation pass would be ideal to make sure there are no discrepancies between backends.

Reply:
Ok... if you don't want to rely on the MIR transformation, you might want to change this signature to return statement instead.

Reply:
I gave this some thought, and I'm not clear on how to make the function return a statement given that the rvalue can be in the context of an assignment, a function call, etc.. I think we should keep the current signature that returns an expression, and we can implement `StatementExpression`-like expressions in Boogie using function calls (via inlined procedures). This would allow us to implement any checks in that function, and then return the expression for the rvalue at the end of the function. What do you think?

Reply:
You would need to pass the place that would be assigned to and generate the assignment here. Both would work.

Reply:
One third option is to return both optional statements and an expr.

Reply:
I went with the third option.